nixops/resources/vault_approle.py

```diff
@@ -0,0 +1,218 @@
+import nixops.util
+import nixops.resources
+import nixops.vault_common
+import json
+import requests
+
+
+class VaultApproleDefinition(nixops.resources.ResourceDefinition):
+    """Definition of a Vault Approle"""
+
+    @classmethod
+    def get_type(cls):
+        return "vault-approle"
+
+    @classmethod
+    def get_resource_type(cls):
+        return "vaultApprole"
+
+    def show_type(self):
+        return "{0}".format(self.get_type())
+
+
+class VaultApproleState(nixops.resources.DiffEngineResourceState):
+    """State of a Vault Approle"""
+
+    state = nixops.util.attr_property("state", nixops.resources.ResourceState.MISSING, int)
+    vault_token = nixops.util.attr_property("vaultToken", None)
+    _reserved_keys = [ "vaultAddress", "roleName", "roleId", "bindSecretId"
+                        "secretIdBoundCidrs", "tokenBoundCidrs", "policies"
+                        "secretIdNumUses", "secretIdTtl", "tokenNumUses"
+                        "tokenMaxTtl", "tokenTtl", "period", "enableLocalSecretIds"
+                        "tokenType", "secretId", "cidrList"
+                      ]
+
+    @classmethod
+    def get_type(cls):
+        return "vault-approle"
+
+    def __init__(self, depl, name, id):
+        nixops.resources.DiffEngineResourceState.__init__(self, depl, name, id)
```

Comment:
I would expect some handlers to be defined here since you're using `DiffEngineResourceState`.
The handler should handle a set of keys e.g `["vaultAddress" "roleName"]` and whenever there is a diff between the config and stored state it will call a function that will handle the realization of the diff. (see https://github.com/NixOS/nixops/blob/master/nixops/resources/vpc_internet_gateway.py#L41 for a simple example)